### PR TITLE
[FSDP2] Changed grad acc test to use data parallel ref model

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126166
* __->__ #126161
* #126070
* #126067

This simplifies the test a bit.

**Context**
Option 1: Ref model is data parallel. Each rank's ref model receives local batch. We manually all-reduce gradients and divide them by world size to match DDP/FSDP semantics.
Option 2: Ref model is not data parallel. Each rank's ref model receives the same global batch. We manually divide the ref model's gradients by world size to match DDP/FSDP semantics. (Note that all ranks have the same ref model and same global batch.)

All of our other unit tests are written following Option 1, which is simpler and a more direct comparison to what our claimed semantics are. This PR switches the gradient accumulation test from being written as following Option 2 to as following Option 1.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 1]
```diff
@@ -797,11 +788,9 @@ def _test_gradient_accumulation(
                 expected_all_gather_count += num_mlps * (num_microbatches - 1)
             self.assertEqual(all_gather_count, expected_all_gather_count)
 
-            # Average the ref model's gradients over the world size to match
-            # data parallel semantics
             for param in ref_model.parameters():
                 if param.grad is not None:
-                    param.grad.div_(self.world_size)
+                    dist.all_reduce(param.grad, op=dist.ReduceOp.AVG)
```
Comment (wanchaol) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:793]: iirc the ref model is a dp model now, whereras the previous approach is more like a `single device` model?

This would incur additional allreduce communications compare to the single device approach though 🤔

**Review by wanchaol — APPROVED** [review]
have one question, either approach lgtm

File: test/distributed/_composable/fsdp/test_fully_shard_training.py (hunk 1 above)
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:793]: It does incur extra all-reduce, but that all-reduce is outside the `CommDebugMode` region (so it does not affect the counts).

**Comment by awgu** [comment]
[inductor / cuda12.1-py3.10-gcc9-sm80 / test (inductor_torchbench_smoketest_perf, 1, 1, linux.gcp.a100)](https://hud.pytorch.org/pr/pytorch/pytorch/126161#24947946551) failure is unrelated
